Client/Frontend/Home/TopSitesPanel.swift

```diff
@@ -0,0 +1,192 @@
+        let c = CGFloat(count)
+        let offset: CGFloat = ToolbarHeight + StatusBarHeight + CGFloat(ContainerHeight)
+
+        if c <= numRows * numCols {
+            let row = floor(Double(c / numCols))
+        	return CGSize(width: width, height: CGFloat(row) * thumbHeight + offset)
+        }
+
+        let h = (c - numRows * numCols) * RowHeight
+        return CGSize(width: width, height: numRows * thumbHeight + h + offset)
+    }
+
+    override func layoutAttributesForElementsInRect(rect: CGRect) -> [AnyObject]? {
+        let start = getIndexAtPosition(rect.origin.y)
+        let end = getIndexAtPosition(rect.origin.y + rect.height)
+
+        var attrs = [UICollectionViewLayoutAttributes]()
+        if start == -1 || end == -1 {
+            return attrs
+        }
+
+        for i in start...end {
+            let indexPath = NSIndexPath(forItem: i, inSection: 0)
+            let attr = layoutAttributesForItemAtIndexPath(indexPath)
+            attrs.append(attr)
+
+            if CGFloat(i) >= (numRows * numCols)-1 {
+                let decoration = layoutAttributesForDecorationViewOfKind(SeperatorKind, atIndexPath: indexPath)
+                attrs.append(decoration)
+            }
+        }
+        return attrs
+    }
+
+    override func layoutAttributesForDecorationViewOfKind(elementKind: String, atIndexPath indexPath: NSIndexPath) -> UICollectionViewLayoutAttributes! {
+        let decoration = UICollectionViewLayoutAttributes(forDecorationViewOfKind: elementKind, withIndexPath: indexPath)
+
+        let h = ((CGFloat(indexPath.item + 1)) - numRows * numCols) * RowHeight
+        decoration.frame = CGRect(x: 0,
+            y: CGFloat(thumbHeight * numRows + h),
+            width: width,
+            height: 1)
+        return decoration
+    }
+
+    override func layoutAttributesForItemAtIndexPath(indexPath: NSIndexPath) -> UICollectionViewLayoutAttributes! {
+        if let cached = attrCache[indexPath.item] {
+            return cached
+        }
+
+        let attr = UICollectionViewLayoutAttributes(forCellWithIndexPath: indexPath)
+
+        let i = CGFloat(indexPath.item)
+        if i < numRows * numCols {
+            let row = floor(Double(i / numCols))
+            let col = i % numCols
+            attr.frame = CGRect(x: CGFloat(thumbWidth * col),
+                y: CGFloat(row) * thumbHeight,
+                width: thumbWidth,
+                height: thumbHeight)
+        } else {
+            let h = CGFloat(i - numRows * numCols) * RowHeight
+            attr.frame = CGRect(x: 0,
+                y: CGFloat(thumbHeight * numRows + h),
+                width: width,
+                height: RowHeight)
+        }
+
+        attrCache[indexPath.item] = attr
+        return attr
+    }
+
+    override func shouldInvalidateLayoutForBoundsChange(newBounds: CGRect) -> Bool {
+        return true
+    }
+}
+
+class TopSitesDataSource: NSObject, UICollectionViewDataSource {
+    var data: Cursor!
```

Comment:
This does not need to be an explicit optional if `init()` takes a non-optional `Cursor`.